src/Functions/array/mapOp.cpp

```diff
@@ -0,0 +1,316 @@
+                val_type = promoted_type;
+                is_float = which_val.isFloat();
+            }
+            else
+            {
+                if (!(k->getNestedType()->equals(*key_type)))
+                    throw Exception(
+                        "All key types in " + getName() + " should be same: " + key_type->getName(),
+                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+                if (is_float != which_val.isFloat())
+                    throw Exception(
+                        "All value types in " + getName() + " should be or float or integer", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+                if (!(promoted_type->equals(*val_type)))
+                {
+                    throw Exception(
+                        "All value types in " + getName() + " should be promotable to " + val_type->getName() + ", got "
+                            + promoted_type->getName(),
+                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+                }
+            }
+
+            if (!res)
+            {
+                res = std::make_shared<DataTypeTuple>(
+                    DataTypes{std::make_shared<DataTypeArray>(k->getNestedType()), std::make_shared<DataTypeArray>(promoted_type)});
+            }
+        }
+
+        return res;
+    }
+
+    template <typename KeyType, typename ValType>
+    void execute2(Block & block, const size_t result, size_t row_count, TupleMaps & args, const DataTypeTuple & res_type)
+    {
+        MutableColumnPtr res_tuple = res_type.createColumn();
+
+        auto * to_tuple = assert_cast<ColumnTuple *>(res_tuple.get());
+        auto & to_keys_arr = assert_cast<ColumnArray &>(to_tuple->getColumn(0));
+        auto & to_keys_data = to_keys_arr.getData();
+        auto & to_keys_offset = to_keys_arr.getOffsets();
+
+        auto & to_vals_arr = assert_cast<ColumnArray &>(to_tuple->getColumn(1));
+        auto & to_vals_data = to_vals_arr.getData();
+
+        size_t res_offset = 0;
+        std::map<KeyType, ValType> summing_map;
+
+        for (size_t i = 0; i < row_count; i++)
+        {
+            [[maybe_unused]] bool first = true;
+            for (auto & arg : args)
+            {
+                size_t offset = 0, len = arg.key_offsets[0];
+
+                if (!arg.is_const)
+                {
+                    offset = i > 0 ? arg.key_offsets[i - 1] : 0;
+                    len = arg.key_offsets[i] - offset;
+
+                    if (arg.val_offsets[i] != arg.key_offsets[i])
+                        throw Exception(
+                            "Key and value array should have same amount of elements", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+                }
+
+                for (size_t j = 0; j < len; j++)
+                {
+                    auto key = arg.key_column.operator[](offset + j).get<KeyType>();
+                    auto value = arg.val_column.operator[](offset + j).get<ValType>();
```

Comment:
But what's the right way of doing this? The `ValType` is the promoted type, and I'm not sure that converting to ColumnVector on promoted type will be right. Also it would require another branch for String columns.

Reply:
Ok, I checked and converting to ColumnVector<promoted_type> definitely doesn't work. And we can't use explicit types here, since value types can be different (for example one Int16 and the second Int32, and we can't use templating for both of the types), so I don't see another solution that using `Field` here.

Reply:
At the end I have optimized working with the keys, but extracting values can be optimized only if I add the requirement where value arrays should have the same type.

Reply:
As you probably (and rightfully) used sumMap as a reference for your implementation, it is totally fine to make use of Field here.
We can get back to this question in a separate task. Your suggested optimization is already a good start.